feat: correct update direction in `for-direction`

fasttime opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->
**Tell us about your environment (`npx eslint --env-info`):**

Node version: v20.5.1
npm version: v9.8.0
Local ESLint version: v8.47.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 22.5.0

**What parser are you using (place an "X" next to just one item)?**

[X] `Default (Espree)`
[ ] `@typescript-eslint/parser`
[ ] `@babel/eslint-parser`
[ ] `vue-eslint-parser`
[ ] `@angular-eslint/template-parser`
[ ] `Other`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
{
    "rules": {
        "for-direction": "error"
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
// False positives:

for (var i = 10; i >= 0; i += this.step);
for (var i = 0; i < 10; i -= (2 - 3));
for (var i = 0; i < 10; i -= foo());
for (var i = 0, n = -1; i < 10; i += -n);
for (var i = max; i >= min; i += 'foo');
for (var i = max; i >= min; i += 0);

// False negatives:

for (var i = 0; i < 10; i += (2 - 3));
for (var i = 0; i < 10; i -= -(-1));
var n = -2; for(var i = 0; i < 10; i += n);
```

[**Repro Link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGZvci1kaXJlY3Rpb246IFwiZXJyb3JcIiAqL1xuXG4vLyBGYWxzZSBwb3NpdGl2ZXM6XG5cbmZvciAodmFyIGkgPSAxMDsgaSA+PSAwOyBpICs9IHRoaXMuc3RlcCk7XG5mb3IgKHZhciBpID0gMDsgaSA8IDEwOyBpIC09ICgyIC0gMykpO1xuZm9yICh2YXIgaSA9IDA7IGkgPCAxMDsgaSAtPSBmb28oKSk7XG5mb3IgKHZhciBpID0gMCwgbiA9IC0xOyBpIDwgMTA7IGkgKz0gLW4pO1xuZm9yICh2YXIgaSA9IG1heDsgaSA+PSBtaW47IGkgKz0gJ2ZvbycpO1xuZm9yICh2YXIgaSA9IG1heDsgaSA+PSBtaW47IGkgKz0gMCk7XG5cbi8vIEZhbHNlIG5lZ2F0aXZlczpcblxuZm9yICh2YXIgaSA9IDA7IGkgPCAxMDsgaSArPSAoMiAtIDMpKTtcbmZvciAodmFyIGkgPSAwOyBpIDwgMTA7IGkgLT0gLSgtMSkpO1xudmFyIG4gPSAtMjsgZm9yKHZhciBpID0gMDsgaSA8IDEwOyBpICs9IG4pO1xuIiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQifX19)

**What did you expect to happen?**

No problem reported for false positives, a problem each for false negatives.

**What actually happened? Please include the actual, raw output from ESLint.**

Errors are reported for false positives, no error for false negatives.

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This PR updates the logic used in the `for-direction` rule to determine the direction (increasing, decreasing, or neither) of the update part of a for-statement that uses the operator `+=` or `-=`.

The updated logic uses full static code analysis instead of the current heuristic to determine the value on the right side of the assignment. This generates both less false positives and less false negatives.

#### Is there anything you'd like reviewers to focus on?

Do the test cases all make sense?

<!-- markdownlint-disable-file MD004 -->

mdjermanovic commented:
The change makes sense to me. As this rule already has a logic to consider some expressions (identifiers) as unknown values and thus not report on them, seems better to use the more advanced utility we have for this.

mdjermanovic reviewed (approved):
LGTM, thanks! I'll leave this open in case someone else wants to review it before merging.

fasttime commented:
Thanks for the suggestion @nzakas, I've added some more examples that match the new behavior. I have also moved the description text up to the overview, since it only describes the problem being addressed, and I've added a line to the Rule Details section to explain what the rule does. Please, check if this is sufficient.

nzakas commented on docs/src/rules/for-direction.md:
```diff
@@ -3,11 +3,11 @@ title: for-direction
 ## Rule Details
 
-A `for` loop with a stop condition that can never be reached, such as one with a counter that moves in the wrong direction, will run infinitely. While there are occasions when an infinite loop is intended, the convention is to construct such loops as `while` loops. More typically, an infinite for loop is a bug.
+This rule forbids for loops where the counter variable increases or decreases in the wrong direction.
```
This read a little weird to me, maybe this?

```suggestion
This rule forbids `for` loops where the counter variable changes in such a way that the stop condition will never be met. For example, if the counter variable is increasing (i.e. `i++`) and the stop condition tests that the counter is greater than zero (`i >=0`) then the loop will never exit.
```
